tools/waveio.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+
+import struct
+import wave
+import numpy as np
+
+def normalize(data, bits):
+    signed_max = get_signed_max_by_bits(bits)
+    return data / signed_max
+
+def get_signed_max_by_bits(bits):
+    return 2 ** (bits - 1) - 1
+
+def quantize(data, bits):
+    np_dtype = np.int16 if bits == 16 else np.int32
+    return (data * get_signed_max_by_bits(bits)).astype(np_dtype)
```

Comment:
See e.g. https://en.wikipedia.org/wiki/Dither#Digital_audio

Reply:
what the numpy does is floor to zero, eg: `4.2 -> 4, 4.8 -> 4, -4.8 -> -4, -4.2 -> -4`, I decide to add 0.5 for the positive part and -0.5 for the negative part, then it will be rounding before transform to intX type. is this OK?

Ps: With currently implementation, I read a wave and write it to a new file, the diff of the two wave in Octave is a zero matrix, but with round, there should be a little difference, I guess,  any explanation?

Reply:
I have just done like y = floor(scale * x + 0.5).  Here's some quick test with Octave:

```
>> x=linspace(-0.1, 0.1, 1001);
>> mean(x)
ans =   -3.1055e-18
>> xqf = floor(x*32768);
>> mean(xqf)
ans = -0.49950
>> xqr = floor(x*32768 + 0.5);
>> mean(xqr)
ans = 0
>> xqnp = floor(x*32768);
>> xqnp(x>0) = floor(x(x>0)*32768 + 0.5);
>> xqnp(x<0) = floor(x(x<0)*32768 - 0.5);
>> mean(xqnp)
ans = -0.49950

```

So I think the +0.5 or -0.5 add creates an offset similarly as just floor.

The rounding should be done if word length is converted to lower or if you do any arithmetic operation like gain. The rounding should not alter samples, at least with 64 bit float that Octave uses.

```
>> x=-32768:32768;
>> xf = x/32768;
>> xr = floor(32768*xf + 0.5);
>> max(abs(x-xr))
ans = 0

```

Reply:
-Inf -------------- 0 --------------- +Inf

Matlab's `round` function
```
Positive part  round -> +Inf      0.2 -> 0    0.8 -> 1
Negative Part, round -> -Inf       -0.2 -> 0  -0.8 -> -1
```

Matlab's `floor` function
```
Positive part  floor -> -Inf      0.2 -> 0    0.8 -> 0
Negative Part, floor-> -Inf       -0.2 -> -1  -0.8 -> -1
```

Numpy's `astype` function
```
Positive part  floor  -> 0      0.2 -> 0    0.8 -> 0
Negative Part, floor-> 0      -0.2 -> 0  -0.8 -> 0
```
@singalsu My previous description with words `floor` or `round` may not accurate. But with above results, things should be much clear. The current implementation with positive + 0.5 and negative -0.5 should behaves just like matlab's round function. so we are good.

> So I think the +0.5 or -0.5 add creates an offset similarly as just floor.

This create offset in matlab, but not in np.astype(), because they don't have same bahavior.

Reply:
If it's like Matlab round() then it's OK. Please add a brief comment how the rounding works with the function that you use!

It's good catch that in Matlab round(x) and floor(x + 0.5) are not same.